src/main/java/com/algolia/search/ConfigBase.java

```diff
@@ -1,44 +1,136 @@
-   *
-   * @param applicationID The Application ID
-   * @param apiKey The API Key
-   * @param executor The Executor service
-   */
-  ConfigBase(String applicationID, String apiKey, ExecutorService executor) {
-    this.applicationID = applicationID;
-    this.apiKey = apiKey;
-    this.executor = executor;
-    this.batchSize = 1000;
-    this.defaultHeaders = new HashMap<>();
-    this.defaultHeaders.put(Defaults.ALGOLIA_APPLICATION_HEADER, getApplicationID());
-    this.defaultHeaders.put(Defaults.ALGOLIA_KEY_HEADER, getApiKey());
-    this.defaultHeaders.put(
-        Defaults.USER_AGENT_HEADER,
-        String.format("Algolia for Java (%s); JVM (%s)", clientVersion, javaVersion));
-    this.defaultHeaders.put(Defaults.ACCEPT_HEADER, Defaults.APPLICATION_JSON);
-    this.defaultHeaders.put(Defaults.ACCEPT_ENCODING_HEADER, Defaults.CONTENT_ENCODING_GZIP);
+  /** Config base builder to ensure the immutability of the configuration. */
+  public abstract static class Builder<T extends Builder<T>> {
+
+    private final String applicationID;
+    private final String apiKey;
+    private final HashMap<String, String> defaultHeaders;
+    private int batchSize;
+    private Integer readTimeOut;
+    private Integer writeTimeOut;
+    private Integer connectTimeOut;
+    private List<StatefulHost> defaultHosts;
+    private List<StatefulHost> customHosts;
+    private final ExecutorService executor;
+
+    /**
+     * Builds a base configuration
+     *
+     * @param applicationID The Algolia Application ID
+     * @param apiKey The API Key: could be Admin API Key or Search API Key
+     * @throws NullPointerException if ApplicationID/ApiKey/Config/Requester is null
+     */
+    public Builder(@Nonnull String applicationID, @Nonnull String apiKey) {
+
+      Objects.requireNonNull(applicationID, "An ApplicationID is required.");
+      Objects.requireNonNull(apiKey, "An API key is required.");
+
+      if (AlgoliaUtils.isEmptyWhiteSpace(applicationID)) {
+        throw new NullPointerException("The ApplicationID can't be empty.");
+      }
+
+      if (AlgoliaUtils.isEmptyWhiteSpace(apiKey)) {
+        throw new NullPointerException("The APIKey can't be empty.");
+      }
+
+      this.applicationID = applicationID;
+      this.apiKey = apiKey;
+      this.batchSize = 1000;
+      this.defaultHeaders = new HashMap<>();
+      this.defaultHeaders.put(Defaults.ALGOLIA_APPLICATION_HEADER, applicationID);
+      this.defaultHeaders.put(Defaults.ALGOLIA_KEY_HEADER, apiKey);
+      this.defaultHeaders.put(
+          Defaults.USER_AGENT_HEADER,
+          String.format("Algolia for Java (%s); JVM (%s)", clientVersion, javaVersion));
+      this.defaultHeaders.put(Defaults.ACCEPT_HEADER, Defaults.APPLICATION_JSON);
+      this.defaultHeaders.put(Defaults.ACCEPT_ENCODING_HEADER, Defaults.CONTENT_ENCODING_GZIP);
+      this.executor = ForkJoinPool.commonPool();
+    }
+
+    /** To prevent unchecked cast warning. */
+    public abstract T getThis();
```

Comment:
OK, actually there is a simpler construction. Tell me what you think:

```java
abstract class ConfigBase {
  public abstract static class Builder<T> {
    // [...]
    public abstract T build();

    public Builder<T> setBatchSize(int batchSize) {
      this.batchSize = batchSize;
      return this;
    }
    // [...]
  }
}
```

And other Builders:
```java
public final class InsightsConfig extends ConfigBase {
  // [...]
  public static class Builder extends ConfigBase.Builder<InsightsConfig> {
    @Override
    public InsightsConfig build() {
      return new InsightsConfig(this);
    }
  }
  // [...]
}
```